From b3254b8bab4fbb391aa8135e191eebc9d50f16de Mon Sep 17 00:00:00 2001
Message-Id: <b3254b8bab4fbb391aa8135e191eebc9d50f16de.1442004380.git.yunwei.zhang@intel.com>
From: Praveen Paneri <praveen.paneri@intel.com>
Date: Fri, 28 Aug 2015 12:02:32 +0530
Subject: [PATCH] SQUASHME! [VPG]: drm/i915: Disable DPST IE in dpms_off

If DPST image enhancements are not disabled during dpms_off, incorrect
image enhancements are applied in subsequent dpms_on call. This
results in visual flicker on the screen if display is turned on before
gfx runtime suspend kicks in. This issue is not seen if device is
allowed to enter (runtime) suspend as IE bit is cleared in the resume
path.
Removing the IE disabling code from gfx suspend-resume path as it will
not be required after the above change.

SQUASHME! - This patch should be squashed into the following existing patch:
Change-Id: I24147618fd125abed61b09f140644d5735f2c386
Author: Daisy Sun <daisy.sun@intel.com>
Date:   Fri Jan 3 16:59:05 2014 -0800
Subject:  MUST_REBASE [VPG]: drm/i915 Enable DPST on Haswell

Change-Id: I2dd5c8b19e84e74b405f771aad25b265f2d08200
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-17475
Signed-off-by: Praveen Paneri <praveen.paneri@intel.com>
---
 drivers/gpu/drm/i915/i915_dpst.c    | 8 ++++++++
 drivers/gpu/drm/i915/i915_suspend.c | 5 -----
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_dpst.c b/drivers/gpu/drm/i915/i915_dpst.c
index 8e42d09..ec954b7 100644
--- a/drivers/gpu/drm/i915/i915_dpst.c
+++ b/drivers/gpu/drm/i915/i915_dpst.c
@@ -510,6 +510,7 @@ void
 i915_dpst_display_off(struct drm_device *dev)
 {
 	struct drm_i915_private *dev_priv = dev->dev_private;
+	u32 blm_hist_ctl;
 
 	/* Check if dpst is user enabled*/
 	if (!dev_priv->dpst.user_enable)
@@ -522,6 +523,13 @@ i915_dpst_display_off(struct drm_device *dev)
 	dev_priv->dpst.display_off = true;
 
 	i915_dpst_disable_hist_interrupt(dev);
+	/*
+	 * Disabling Image Enhancement bit during display off to avoid
+	 * applying incorrect IE upon resume
+	 */
+	blm_hist_ctl = I915_READ(dev_priv->dpst.reg.blm_hist_ctl);
+	blm_hist_ctl &= ~(dev_priv->dpst.ie_mod_table_enable);
+	I915_WRITE(dev_priv->dpst.reg.blm_hist_ctl, blm_hist_ctl);
 	mutex_unlock(&dev_priv->dpst.ioctl_lock);
 
 	/* Send a fake signal to user, so that the user can be notified
diff --git a/drivers/gpu/drm/i915/i915_suspend.c b/drivers/gpu/drm/i915/i915_suspend.c
index ddd0c86..ba18d58 100644
--- a/drivers/gpu/drm/i915/i915_suspend.c
+++ b/drivers/gpu/drm/i915/i915_suspend.c
@@ -197,11 +197,6 @@ void i915_save_dpst_regs(struct drm_i915_private *dev_priv)
 
 	dev_priv->regfile.saveBLM_HIST_GUARD = I915_READ(dev_priv->dpst.reg.blm_hist_guard);
 	dev_priv->regfile.saveBLM_HIST_CTL = I915_READ(dev_priv->dpst.reg.blm_hist_ctl);
-
-	/* Disable image enhancement table so we do not apply invalid
-	 * data when we resume */
-	dev_priv->regfile.saveBLM_HIST_CTL &=
-		~(dev_priv->dpst.ie_mod_table_enable);
 }
 
 void i915_restore_dpst_regs(struct drm_i915_private *dev_priv)
-- 
1.9.1

